stdlib/public/core/KeyPathIterable.swift

```diff
@@ -0,0 +1,115 @@
+//===-- KeyPathIterable.swift ---------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the KeyPathIterable protocol.
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// KeyPathIterable
+//===----------------------------------------------------------------------===//
+
+/// An implementation detail of `KeyPathIterable`; do not use this protocol
+/// directly.
+public protocol _KeyPathIterableBase {
+  var _allKeyPathsTypeErased: [AnyKeyPath] { get }
+  var _recursivelyAllKeyPathsTypeErased: [AnyKeyPath] { get }
+}
+
+/// A type whose values provides custom key paths to properties or elements.
+public protocol KeyPathIterable : _KeyPathIterableBase {
+  /// A type that can represent a collection of all key paths of this type.
+  associatedtype AllKeyPaths : Collection
+    where AllKeyPaths.Element == PartialKeyPath<Self>
+
+  /// A collection of all custom key paths of this value.
+  var allKeyPaths: AllKeyPaths { get }
+}
+
+public extension KeyPathIterable {
+  /// An array of all custom key paths of this value and any custom key paths
+  /// nested within each of what this value's key paths refers to.
+  var recursivelyAllKeyPaths: [PartialKeyPath<Self>] {
```

Comment:
The naming of `recursivelyAllKeyPaths` and the doc comment have room for improvement.

Reply:
`nestedKeyPaths`?

Reply:
Thanks for the suggestion! I feel `nestedKeyPaths` is not an ideal name for this property because it also contains top-level `allKeyPaths`. `nestedKeyPaths` sounds like it *doesn't* contain top-level properties (or at least leaves it ambiguous).

---

We considered `nestedKeyPaths` as not just a different name, but a different design option (where `allNestedKeyPaths` refers to `recursivelyAllKeyPaths` excluding top-level `allKeyPaths`, and `recursivelyAllKeyPaths` would be defined as `allKeyPaths + allNestedKeyPaths`).

We decided not to implement that option though, since we couldn't think of many use cases for "just recursively nested non-top-level key paths".